bin/executables/metrics.rb

```diff
@@ -0,0 +1,157 @@
+    #  are queried from the vCenter database and can take significantly longer to gather (> 5 minutes),
+    #  depending on the performance of the vCenter host. Having two threads keeps the metrics collector from
+    #  becoming permanently backlogged.
+    def initialize(scheduler)
+      logger.info 'Initializing Metrics gathering process'
+      @scheduler = scheduler
+    end
+
+    def execute
+      logger.info '- Loading Metrics from Vsphere handler'
+
+      begin
+        main_thread
+        [current_thread, backlog_thread].map(&:join)
+      rescue StandardError => e
+        logger.fatal "Encountered unhandled exception: #{e.message}."
+        logger.debug e.backtrace
+        @scheduler.shutdown
+        exit(1)
+      end
+      logger.info '- Metrics Loaded from Vsphere handler'
+    end
+
+    private
+
+    def collector_creator
+      MetricsCollector.new
+    rescue StandardError => e
+      logger.fatal "Unable to start metrics collection: #{e.message}"
+      logger.debug e.backtrace
+      @scheduler.shutdown
+      exit(1)
+    end
+
+    def current_collector
+      @current_collector ||= collector_creator
+    end
+
+    def backlog_collector
+      @backlog_collector ||= collector_creator
+    end
+
+    def current_queue
+      @current_queue ||= Queue.new
+    end
+
+    def backlog_queue
+      @backlog_queue ||= Queue.new
+    end
+
+    def main_thread
+      if current_queue.empty? || backlog_queue.empty?
+        # Iterate over timestamps (from oldest to newest).
+        #  Queue is not being used here in the usual sense - its main concern is to provide a simple,
+        #  thread-safe variable. We only enqueue when one is empty to keep the number of InventoriedTimestamps
+        #  in the state of 'queued_for_metering' or 'metering' to a minimum. This reduces the amount
+        #  of readings the missing readings handler will have to (needlessly) map/reduce as part of its job.
+        inventoried_timestamps =  InventoriedTimestamp.where(record_status: 'inventoried',
+                                   locked: false,
+                                   :inventory_at.lte => 6.minutes.ago,
+                                   :inventory_at.gte => 23.hours.ago)
+            .asc(:inventory_at)
+        logger.info "INVENTORIED TIMESTAMPS => \n"
+        inventoried_timestamps.map do |it|
+          logger.info "#{it.inspect} \n"
+        end
+        inventoried_timestamps.each do |it|
+          # 55 minutes used to provide a bit of "wiggle room", and hopefully avoid attempting
+          #  collection directly on the hour, when stats may be getting purged from ESX
+          if it.inventory_at >= 55.minutes.ago && current_queue.empty?
+            begin
+              it.with_lock do
+                it.update_attributes(record_status: 'queued_for_metering', locked: true)
+                logger.info "IT C=> #{it.inspect}\n"
```

Comment:
Maybe we could include something easier to find later (something like ----- or similar)